File: frontend/src/lib/phoneUtils.ts
/**
 * Format phone number for display (US format)
 * @param phone - Raw phone number string
 * @returns Formatted phone number or original if invalid
 */
export function formatPhoneNumber(phone: string): string {
  if (!phone) return '';
  
  // Remove all non-numeric characters
  const cleaned = phone.replace(/\D/g, '');
  
  // Handle different lengths
  if (cleaned.length === 10) {
    // (XXX) XXX-XXXX
    return `(${cleaned.slice(0, 3)}) ${cleaned.slice(3, 6)}-${cleaned.slice(6)}`;
  } else if (cleaned.length === 11 && cleaned.startsWith('1')) {
    // +1 (XXX) XXX-XXXX
    return `+1 (${cleaned.slice(1, 4)}) ${cleaned.slice(4, 7)}-${cleaned.slice(7)}`;
  } else if (cleaned.length === 7) {
    // XXX-XXXX (local number)
    return `${cleaned.slice(0, 3)}-${cleaned.slice(3)}`;
  }
  
  // Return original if we can't format it
  return phone;
}

/**
 * Clean phone number for storage (remove formatting)
 * @param phone - Formatted phone number
 * @returns Clean phone number with only digits and + for international
 */
export function cleanPhoneNumber(phone: string): string {
  if (!phone) return '';
  
  const cleaned = phone.replace(/\D/g, '');
  
  // Add +1 prefix for US numbers if 10 digits
  if (cleaned.length === 10) {
    return `+1${cleaned}`;
  } else if (cleaned.length === 11 && cleaned.startsWith('1')) {
    return `+${cleaned}`;
  }
  
  return cleaned;
}

/**
 * Validate phone number format
 * @param phone - Phone number to validate
 * @returns true if valid US phone number
 */
export function isValidPhoneNumber(phone: string): boolean {
  if (!phone) return false;
  
  const cleaned = phone.replace(/\D/g, '');
  
  // Valid US phone numbers: 10 digits or 11 digits starting with 1
  return cleaned.length === 10 || (cleaned.length === 11 && cleaned.startsWith('1'));
}

/**
 * Format phone number as user types (for input fields)
 * @param value - Current input value
 * @returns Formatted value with cursor position info
 */
export function formatPhoneInput(value: string): { formatted: string; cursorOffset: number } {
  const cleaned = value.replace(/\D/g, '');
  let formatted = '';
  let cursorOffset = 0;
  
  if (cleaned.length >= 6) {
    formatted = `(${cleaned.slice(0, 3)}) ${cleaned.slice(3, 6)}-${cleaned.slice(6, 10)}`;
  } else if (cleaned.length >= 3) {
    formatted = `(${cleaned.slice(0, 3)}) ${cleaned.slice(3)}`;
  } else {
    formatted = cleaned;
  }
  
  return { formatted, cursorOffset };
}

/**
 * React hook for phone number input with real-time formatting
 */
import { useState, useCallback } from 'react';

export function usePhoneInput(initialValue: string = '') {
  const [value, setValue] = useState(formatPhoneNumber(initialValue));
  const [rawValue, setRawValue] = useState(cleanPhoneNumber(initialValue));
  
  const handleChange = useCallback((inputValue: string) => {
    const { formatted } = formatPhoneInput(inputValue);
    setValue(formatted);
    setRawValue(cleanPhoneNumber(formatted));
  }, []);
  
  const isValid = isValidPhoneNumber(value);
  
  return {
    value,
    rawValue, // Use this for API calls
    handleChange,
    isValid,
    reset: () => {
      setValue('');
      setRawValue('');
    }
  };
}

File: backend/app/utils/phone_utils.py
import re
from typing import Optional

def clean_phone_number(phone: str) -> Optional[str]:
    """
    Clean and standardize phone number for storage.
    Stores in E.164 format: +1XXXXXXXXXX for US numbers
    
    Args:
        phone: Raw phone input from user
        
    Returns:
        Cleaned phone number or None if invalid
    """
    if not phone or not isinstance(phone, str):
        return None
    
    # Remove all non-numeric characters
    cleaned = re.sub(r'[^\d]', '', phone.strip())
    
    if not cleaned:
        return None
    
    # Handle different US phone number formats
    if len(cleaned) == 10:
        # Standard US 10-digit number
        return f"+1{cleaned}"
    elif len(cleaned) == 11 and cleaned.startswith('1'):
        # 11-digit starting with 1
        return f"+{cleaned}"
    elif len(cleaned) == 7:
        # 7-digit local number - return as-is for now
        # You might want to add area code logic here
        return cleaned
    
    # For international numbers or other formats, return as-is if reasonable length
    if 7 <= len(cleaned) <= 15:
        return f"+{cleaned}"
    
    return None

def format_phone_display(phone: str) -> str:
    """
    Format phone number for display in UI.
    
    Args:
        phone: Stored phone number (should be in +1XXXXXXXXXX format)
        
    Returns:
        Formatted phone number for display
    """
    if not phone:
        return ""
    
    # Remove + and any non-digits
    cleaned = re.sub(r'[^\d]', '', phone)
    
    if len(cleaned) == 11 and cleaned.startswith('1'):
        # US number: +1 (XXX) XXX-XXXX
        return f"({cleaned[1:4]}) {cleaned[4:7]}-{cleaned[7:]}"
    elif len(cleaned) == 10:
        # US number without country code: (XXX) XXX-XXXX
        return f"({cleaned[:3]}) {cleaned[3:6]}-{cleaned[6:]}"
    elif len(cleaned) == 7:
        # Local number: XXX-XXXX
        return f"{cleaned[:3]}-{cleaned[3:]}"
    
    # For other formats, return the original
    return phone

def validate_phone_number(phone: str) -> bool:
    """
    Validate if phone number is in acceptable format.
    
    Args:
        phone: Phone number to validate
        
    Returns:
        True if valid, False otherwise
    """
    if not phone:
        return False
    
    cleaned = re.sub(r'[^\d]', '', phone)
    
    # Accept US phone numbers (10 or 11 digits) or international (7-15 digits)
    if len(cleaned) == 10:
        return True
    elif len(cleaned) == 11 and cleaned.startswith('1'):
        return True
    elif 7 <= len(cleaned) <= 15:
        return True
    
    return False